dotCMS/src/integration-test/java/com/dotcms/security/secret/ServiceIntegrationAPIImplTest.java

```diff
@@ -0,0 +1,172 @@
+                .build();
+
+        final ServiceIntegrationAPI api = APILocator.getServiceIntegrationAPI();
+
+        api.saveServiceSecrets(bean1Host1, host1, admin);
+        api.saveServiceSecrets(bean2Host1, host1, admin);
+        api.saveServiceSecrets(bean3Host1, host1, admin);
+        api.saveServiceSecrets(bean1Host2, host2, admin);
+
+        final Map<String,List<String>> serviceKeysByHost = api.serviceKeysByHost();
+        assertNotNull(serviceKeysByHost.get(host1.getIdentifier()));
+        assertNotNull(serviceKeysByHost.get(host2.getIdentifier()));
+
+        assertEquals(3, api.listServiceKeys(host1).size());
+        assertEquals(1, api.listServiceKeys(host2).size());
+
+        assertEquals(3, serviceKeysByHost.get(host1.getIdentifier()).size());
+        assertEquals(1, serviceKeysByHost.get(host2.getIdentifier()).size());
+
+        assertTrue(serviceKeysByHost.get(host1.getIdentifier()).contains("serviceKey-1-Host-1".toLowerCase()));
+        assertTrue(serviceKeysByHost.get(host1.getIdentifier()).contains("serviceKey-2-Host-1".toLowerCase()));
+        assertTrue(serviceKeysByHost.get(host1.getIdentifier()).contains("serviceKey-3-Host-1".toLowerCase()));
+        assertTrue(serviceKeysByHost.get(host2.getIdentifier()).contains("serviceKey-1-Host-2".toLowerCase()));
+
+    }
+
+    @Test
+    public void Test_Lookup_Fallback() throws DotDataException, DotSecurityException {
+        final ServiceIntegrationAPI api = APILocator.getServiceIntegrationAPI();
+        final User admin = TestUserUtils.getAdminUser();
+        final Host systemHost = APILocator.systemHost();
+        final Host host1 = new SiteDataGen().nextPersisted();
+
+        final ServiceSecrets.Builder builder1 = new ServiceSecrets.Builder();
+
+        final ServiceSecrets beanSystemHost1 = builder1.withServiceKey("serviceKey-1-Any-Host")
+                .withHiddenSecret("fallback:test:secret1", "sec1")
+                .withSecret("fallback:test", true)
+                .build();
+
+        api.saveServiceSecrets(beanSystemHost1, systemHost, admin);
+
+        final Optional<ServiceSecrets> serviceSecretsOptional = api.getSecretForService("serviceKey-1-Any-Host", host1, admin);
+        assertTrue(serviceSecretsOptional.isPresent());
+        final ServiceSecrets recoveredBean = serviceSecretsOptional.get();
+        assertEquals("serviceKey-1-Any-Host", recoveredBean.getServiceKey());
+        assertTrue(recoveredBean.getSecrets().containsKey("fallback:test"));
+        assertTrue(recoveredBean.getSecrets().containsKey("fallback:test:secret1"));
+    }
+
+    @Test(expected = DotSecurityException.class)
+    public void Test_Non_Admin_User_Read_Attempt() throws DotDataException, DotSecurityException {
+        final User nonAdmin = TestUserUtils.getChrisPublisherUser();
+        final ServiceIntegrationAPI api = APILocator.getServiceIntegrationAPI();
+        ServiceSecrets.Builder builder = new ServiceSecrets.Builder();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Local variable 'builder' could be declared final](https://app.codacy.com/app/dotCMS/core/pullRequest?prid=4623512)